### ISSUE TITLE
422 with "value is not a valid list" when a dependency override needs a list argument (which it won't get)

### ISSUE THREAD

**Issue by admackin** [tiangolo/fastapi#4704]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
## requirements.txt

fastapi==0.75.0
uvicorn==0.17.5
pytest==5.2
requests==2.24.0

## main.py

from fastapi import FastAPI, Request, Depends
from pydantic import BaseModel

app = FastAPI()


class Empty(BaseModel):
    pass


def get_depend():
    return None


@app.post("/empty")
def empty(empty: Empty, depend=Depends(get_depend)):
    pass

## test.py

from main import app, get_depend
from starlette.testclient import TestClient

def needs_list_arg(foo: list):  # arg won't get passed
    return None

app.dependency_overrides[get_depend] = needs_list_arg

def test_call_empty():
    response = TestClient(app).post("/empty", json={})
    print(response.content)
    assert response.status_code == 200
```

### Description

* run provided test case (eg using `pytest`), which just (correctly) provides an empty body to the API
* **expected** since there is a dependency override for `get_depend` to `needs_list_arg` which needs a list argument, it should probably fail early with an error of some kind when setting up or using the dependencies early
* **observed**:
```
============================================= FAILURES =============================================
_________________________________________ test_call_empty __________________________________________

    def test_call_empty():
        response = TestClient(app).post("/empty", json={})
        print(response.content)
>       assert response.status_code == 200
E       assert 422 == 200
E        +  where 422 = <Response [422]>.status_code

test.py:16: AssertionError
--------------------------------------- Captured stdout call ---------------------------------------
b'{"detail":[{"loc":["body"],"msg":"value is not a valid list","type":"type_error.list"}]}'
======================================== 1 failed in 0.17s =========================================
```
but there is nothing wrong with the body! this took me a long time to track down, because I had an override set up incorrectly.

changing the arg type declaration in `needs_list_arg` to a `dict` interestingly removes the problem, so it sounds like the way pydantic is used might be related (but I'm fairly confident it's in FastAPI).

Removing the override also stops the issue of course.

### Operating System

macOS

### FastAPI Version

0.75.0

### Python Version

3.8.9

### Additional Context

sounds similar to but is distinct from #2356

**Comment by invokermain** [comment]
Yeah I agree that this area of FastAPI needs some work. Dependency injection is supposed to make testing easier, but obscure 422 errors actually make it a lot harder to debug a failing test. How are you supposed to pass a Mock object as a dependency override?

I feel like overridden dependencies shouldn't be validated with Pydantic?
